### PR TITLE
Fix very rare test_fast_follow failure.

### PR DESCRIPTION
**Pull request by potiuk**
There is an occasional (VERY RARE) case where this test failed when lambda was without delay
This problem resulted with psutil.wait() behaviour which returns None in case the process
that we wait for already exited. Psutil does not have any more information about the process
Nor cannot get the return code so it returns None which results in
`State of this instance has been externally set to success. Terminating instance.` from heartbeat
this breaks the flow of processing for that test. That's why we added small timeout to decrease
the likelihood of the race condition.

The error manifested itself with:

```
  tests/jobs/test_local_task_job.py:704:
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

  self = <tests.jobs.test_local_task_job.TestLocalTaskJob object at 0x7f9ba8376750>
  dag_run = <DagRun test_dagrun_fast_follow @ 2021-10-13 05:32:21.521590+00:00: test_dagrun_fast_follow, externally triggered: False>
  ti_state_mapping = {'A': <TaskInstanceState.SUCCESS: 'success'>, 'B': <TaskInstanceState.SCHEDULED: 'scheduled'>, 'C': None}
  error_message = 'A -> B -> C, with fast-follow ON when A runs, B should be QUEUED. Same for B and C.'

      def validate_ti_states(self, dag_run, ti_state_mapping, error_message):
          for task_id, expected_state in ti_state_mapping.items():
              task_instance = dag_run.get_task_instance(task_id=task_id)
              task_instance.refresh_from_db()
  >           assert task_instance.state == expected_state, error_message
  E           AssertionError: A -> B -> C, with fast-follow ON when A runs, B should be QUEUED. Same for B and C.
  E           assert None == <TaskInstanceState.SCHEDULED: 'scheduled'>
  E            +  where None = <TaskInstance: test_dagrun_fast_follow.B test_dagrun_fast_follow [None]>.state

  tests/jobs/test_local_task_job.py:94: AssertionError
  ---------------------------- Captured stdout setup -----------------------------
  [2021-10-13 05:32:21,463] {dagbag.py:500} INFO - Filling up the DagBag from /dev/null
  ------------------------------ Captured log setup ------------------------------
  INFO     airflow.models.dagbag.DagBag:dagbag.py:500 Filling up the DagBag from /dev/null
  ----------------------------- Captured stdout call -----------------------------
  [2021-10-13 05:32:21,499] {dag.py:2389} INFO - Sync 1 DAGs
  [2021-10-13 05:32:21,504] {dag.py:2408} INFO - Creating ORM DAG for test_dagrun_fast_follow
  [2021-10-13 05:32:21,509] {dag.py:2914} INFO - Setting next_dagrun for test_dagrun_fast_follow to 2016-01-01T00:00:00+00:00
  Running <TaskInstance: test_dagrun_fast_follow.A test_dagrun_fast_follow [running]> on host f6d28748e57e
  [2021-10-13 05:32:22,784] {local_task_job.py:212} WARNING - State of this instance has been externally set to success. Terminating instance.
  ------------------------------ Captured log call -------------------------------
  INFO     airflow.models.dag:dag.py:2389 Sync 1 DAGs
  INFO     airflow.models.dag:dag.py:2408 Creating ORM DAG for test_dagrun_fast_follow
  INFO     airflow.models.dag:dag.py:2914 Setting next_dagrun for test_dagrun_fast_follow to 2016-01-01T00:00:00+00:00
  WARNING  airflow.jobs.local_task_job.LocalTaskJob:local_task_job.py:212 State of this instance has been externally set to success. Terminating instance.
``

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
Hey @ash, @ephraimbuddy @uranusjr - while looking at some test failures I found that very interesting one, which potenatilly **might** occur in real-life production, but I am not sure how to handle it there (but I think the likelihood of that is very, very slim). Probably we can ignore it.

The error in our tests : https://github.com/apache/airflow/runs/3878745891?check_suite_focus=true#step:6:4566 occured only because out python callables were lambdas returning True, without any delay. See the code in the commit message but it manifested itself with heartbeat callback printing "State of this instance has been externally set to success. Terminating instance.". 

I looked at the logs and the code, analyzed all the possible paths and I deduced that the only place the heartbeat callback can be called from is this part in the local_task_job (return_code would have to return None without timeout):

```
                return_code = self.task_runner.return_code(timeout=max_wait_time)
                if return_code is not None:
                    self.handle_task_exit(return_code)
                    return

                self.heartbeat()
```

And when I looked a return_code() method, the only place where it could return None was if the process.wait() returns None itself (but we neew the process should return 0 becasue it was a very quick lambda). So I expected it to return 0.

```
    def return_code(self, timeout: int = 0) -> Optional[int]:
        # We call this multiple times, but we can only wait on the process once
        if self._rc is not None or not self.process:
            return self._rc

        try:
            self._rc = self.process.wait(timeout=timeout)
            self.process = None
        except psutil.TimeoutExpired:
            pass

        return self._rc
```

I looked even closer and (to may great surprise!) I found that psutil.wait() CAN return None even if the child process return 0. This will happen when the process is already terminated at the moment when you call wait() !!!!! (VERY SURPRISING but this is how `psutil.Processwait()` works). 

https://psutil.readthedocs.io/en/latest/#psutil.Process.wait

> If PID does not exist return None immediately.

We are creating the process object using the PID when forking (`return psutil.Process(pid)`), and beforeyou call process.wait() is called, the return value of the child process is not stored anywhere. Unde the hood the `wait` calls `_psposix.wait()` and there you can find the same info: `If PID does not exist at all return None immediately.`,

I think the chance that in "reality" the forked proces will terminate before we start waiting for it is VERY SLIM. And I cannot think of a better way of waiting for the forked process heere. So I think just adding timeout to test should fix the occasional test failures and there should be no more action - but I would like you to take a look and confirm my findings.

**Comment by ashb** [comment]
@potiuk my understanding is that process not existing is different from the process has exited.

An exited process still exists until you call one of the POSIX wait functions.

The only way it can not exist is if something else called wait, or the pid is wrong

**Comment by potiuk** [comment]
Also this one (https://stackoverflow.com/questions/20624957/python-lib-psutil-how-to-get-exit-status-with-wait) says:

> On top of that, once the parent calls wait (which Popen.communicate does), the process—and its entry in the system process table, with its retcode—may not even exist anymore.

I think the .wait() works reliably when we are using Popen to open a process, but I think the problem is that when using `fork  + psutil.Process(pid) and then process.wait()` - there is this really, really short period of time when if the forked process terminates, we might not be able to get it's return code.

And yeah, I am equally surprised by that as you are because I was pretty sure it work differently.. But I also literally cannot explain how else the test could fail in the way it did.

**Comment by potiuk** [comment]
Loudly thinking again: other reasons it could happen: 
* double fork() and exit from `middle` parent (init will wait())
* SIGCHILD ignored by parent process (kernel will dispose the process table).

File: tests/jobs/test_local_task_job.py [hunk 1]
```diff
@@ -656,10 +656,19 @@ def test_fast_follow(
         self, conf, dependencies, init_state, first_run_state, second_run_state, error_message, dag_maker
     ):
 
+        # There is an occasional (VERY RARE) case where this test failed when lambda was without delay
```
Comment (uranusjr) [inline tests/jobs/test_local_task_job.py:659]: Is this sentence incomplete?

**Pull request closed without merging**